cmd/goal/clerk.go

```diff
@@ -548,3 +706,187 @@ var splitCmd = &cobra.Command{
+	Use:   "compile",
+	Short: "compile a contract program",
+	Long:  "compile a contract program, report its address",
+	Run: func(cmd *cobra.Command, args []string) {
+		for _, fname := range args {
+			if disassesmble {
+				disassembleFile(fname, outFilename)
+				continue
+			}
+			program := assembleFile(fname)
+			outblob := program
+			outname := outFilename
+			if outname == "" {
+				outname = fmt.Sprintf("%s.tok", fname)
+			}
+			if signProgram {
+				dataDir := ensureSingleDataDir()
+				accountList := makeAccountsList(dataDir)
+				client := ensureKmdClient(dataDir)
+				wh, pw := ensureWalletHandleMaybePassword(dataDir, walletName, true)
+
+				// Check if from was specified, else use default
+				if account == "" {
+					account = accountList.getDefaultAccount()
+					if account == "" {
+						reportErrorln("no default account set. set one with 'goal account -f' or specify an account with '-a'.")
+					}
+					fmt.Printf("will use default account: %v\n", account)
+				}
+				signingAddressResolved := accountList.getAddressByName(account)
+
+				signature, err := client.SignProgramWithWallet(wh, pw, signingAddressResolved, program)
+				if err != nil {
+					reportErrorf(errorSigningTX, err)
+				}
+				ls := transactions.LogicSig{Logic: program, Sig: signature}
+				outblob = protocol.Encode(ls)
+			}
+			if !noProgramOutput {
+				fout, err := os.Create(outname)
+				if err != nil {
+					reportErrorf("%s: %s\n", outname, err)
+				}
+				_, err = fout.Write(outblob)
+				if err != nil {
+					reportErrorf("%s: %s\n", outname, err)
+				}
+				err = fout.Close()
+				if err != nil {
+					reportErrorf("%s: %s\n", outname, err)
+				}
+			}
+			if !signProgram {
+				pd := transactions.HashProgram(program)
+				addr := basics.Address(pd)
+				fmt.Printf("%s: %s\n", fname, addr.String())
+			}
+		}
+	},
+}
+
+var dryrunCmd = &cobra.Command{
+	Use:   "dryrun",
+	Short: "test a program offline",
+	Long:  "test a program offline under various conditions and verbosity",
+	Run: func(cmd *cobra.Command, args []string) {
+		seed, err := base64.StdEncoding.DecodeString(seedBase64)
+		if err != nil {
+			reportErrorf("invalid seed: %s", err)
+		}
+		data, err := readFile(txFilename)
+		if err != nil {
+			reportErrorf(fileReadError, txFilename, err)
+		}
+		dec := protocol.NewDecoderBytes(data)
+		stxns := make([]transactions.SignedTxn, 0, 10)
+		for {
+			var txn transactions.SignedTxn
+			err = dec.Decode(&txn)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				reportErrorf(txDecodeError, txFilename, err)
+			}
+			stxns = append(stxns, txn)
+		}
+		txgroup := make([]transactions.SignedTxnWithAD, len(stxns))
+		for i, st := range stxns {
+			txgroup[i].SignedTxn = st
+		}
+		if timeStamp == 0 {
+			timeStamp = time.Now().Unix()
+		}
+		block := bookkeeping.Block{}
+		block.BlockHeader.Round = basics.Round(round)
+		block.BlockHeader.TimeStamp = timeStamp
+		proto := config.Consensus[protocol.ConsensusCurrentVersion]
```

Comment:
If `ConsensusCurrentVersion` doesn't support TEAL contracts, then this command won't work.